### ISSUE TITLE
mavericks

### ISSUE THREAD

**Issue by flying-sheep** [eddelbuettel/drat#20]
on newer OSX (currently mavericks and yosemite) versions, the `pkgType` is `mac.binary.mavericks`, and the path `file.path("bin", "macosx", "mavericks", "contrib", rversion)`

[you don’t handle that](https://github.com/eddelbuettel/drat/blob/master/R/insertPackage.R#L164)

**Comment by eddelbuettel** [comment]
Well it so happens that just last evening I added the following to the [(draft) Drat FAQ](https://github.com/eddelbuettel/drat/blob/master/vignettes/DratFAQ.Rmd) vignette's section:

> However, not all functions have been extended, and support for OS X has not been added either. Contributions would be more than welcome.

I do not have an OS X machine.   @JanSchulz did really well in adding the windows support he needed, adding OS X paths and binaries mostly needs some testing.  Can you help?

**Comment by flying-sheep** [comment]
sadly i can’t really, as i neither have an OSX machine. maybe #21 is sufficient…

**Comment by JanSchulz** [comment]
The easiest way would be to use `.Platform$pkgType` of the R session which runs drat and if that's `"mac.binary.mavericks"` use the mavericks path. This is similar to the way the R version is currently handled.

The more right fix would be to use the information in the package DESCRIPTION (insight the *.tgz) to figure out the R version and operating system version which built the package instead of using the values from the R session which runs drat. But that would mean looking into the *.tgz (untar to a tempfile?) and parsing the `Built:` line in the DESCRIPTION file (R Version + kernel version on MacOSX).

**Comment by eddelbuettel** [comment]
Yes, we can look into the tarfile. CRANberries does that too in tempfiles.  

Do you have access to an OS X box?

**Comment by JanSchulz** [comment]
sort of... It's not mine, but I can probably run a few tests on a mac...

Is there any source for CRANberries available? I looked at your github repos, but didn't find it there. Otherwise I think I can get by by using the code from https://github.com/wch/r-source/blob/R-3-2-branch/src/library/tools/R/packages.R#L182 More interesting will be to figure out the version info from the `Built:` line...

So, more or less:
* read the DESCRIPTION file from tgz and zip files
* get the major.minor R version and the os version (only needed for Mac OS X packages -> probably enough to use a regex for versions which are snow leopard and assume mavericks for all else?) from the `Built:` line in DESCRIPTION
* build the to-be-copied-to-path by looking at these infos

**Comment by eddelbuettel** [comment]
I'll send you a snippet off-line.  CRANberries should be on GitHub but I never pushed it because to config to actually _run_ it is tied to my machine and the server / directory layout.  But with RcppTOML it is trivial to farm that off to a config file so I'll just do that....

**Comment by flying-sheep** [comment]
well, the `Built` line reads like this when there’s actually stuff being compiled:

* Built: R 3.2.0; x86_64-apple-darwin13.4.0; 2015-04-27 20:26:59 UTC; unix

and like this if it’s a plain R package:

* Built: R 3.2.0; ; 2015-04-27 20:26:59 UTC; unix

---

darwin 13 is used in mavericks, 14 in yosemite, … so we could parse for that number and use anything >= 13 to designate a “bin.macosx.mavericks” type package.

packages without that fragment are platform-independent anyway, so… install them in both directories?

**Comment by JanSchulz** [comment]
So, my current understanding of `install.packages` (based on https://github.com/IRkernel/IRkernel/issues/50#issuecomment-101594531 and the following comments):

On mac, the third component from "type" (either from  `.Platform$pkgType` (default for type) or by specifying  `type = 'mac.binary.mavericks'` or `type = 'mac.binary'`) will be used to add to the path where it looks for mac binaries.

So the mac binary packages for an up to date system should be found in `bin/macosx/mavericks/contrib/3.y`, but only adding the package to that path will fail if a user uses an explicit `type = 'mac.binary'. -> This problem should go into some readme...

I'm not sure if there is any harm in adding a package to both places (symlink or copy) but I suspect that there must be a reason for adding the third component (but I couldn't find one yet :-( )...

**Comment by flying-sheep** [comment]
binary incompatibility. packages that aren’t pure R but have compiled parts aren’t guaranteed to work when compiled on mavericks or higher and used on a lower version.

so all packages that aren’t compiled can be freely put into both directories, only rzmq has (optimally) to be compiled on both a modern OSX and an old one.

**Comment by eddelbuettel** [comment]
Right.  I briefly looked into this prepareing drat 0.0.4 which is now out. 

As to whether we have to install the file "twice" or not:  can we maybe look at some existing repos and see if the architecture-independent packages (ie a .tgz without compiled code, such as drat) goes to both places?  I see

```
http://cran.rstudio.com/bin/macosx/contrib/3.2/drat_0.0.3.tgz
http://cran.rstudio.com/bin/macosx/mavericks/contrib/3.2/drat_0.0.3.tgz
```

So do we really need to install twice?

**Comment by flying-sheep** [comment]
well, all proper filesystems have symlinks, but github pages doesn’t support them.

so either we bug GH to fix that shit or we copy files.

**Comment by mbjones** [comment]
@JanSchulz Any update on adding the Built: parsing for Mavericks/Yosemite?  I just ran into this problem locally as well with drat_0.0.4. See sessionInfo output below for details. When I ran `drat::insertPackage()` it placed the package in `bin/macosx/contrib/3.2/`, but when I use `drat::addRepo()` and then `install.packages("foo")`, it looks for the package in `bin/macosx/mavericks/contrib/3.2`. Is this what you expected to see on Yosemite, with the mavericks in the path?  I could test any proposed solutions on MacOSX if needed.

```r
> sessionInfo()
R version 3.2.0 (2015-04-16)
Platform: x86_64-apple-darwin13.4.0 (64-bit)
Running under: OS X 10.10.3 (Yosemite)

locale:
[1] en_US.UTF-8/en_US.UTF-8/en_US.UTF-8/C/en_US.UTF-8/en_US.UTF-8

attached base packages:
[1] stats     graphics  grDevices utils     datasets  methods   base     

loaded via a namespace (and not attached):
[1] drat_0.0.4  tools_3.2.0
```

**Comment by eddelbuettel** [comment]
@mbjones  Thanks for the followup.  I *think* based on earlier analysis that we can figure this based on the binary packages. We "just" need to look inside the .zip (for Windoze) or .tgz (for Macheads) and parse the line of DESCRIPTION.  I simply haven't gotten to that yet...

**Comment by mbjones** [comment]
@eddelbuettel I started looking into this. It looks straightforward, as the package types seem to be limited to "mac.binary" and "mac.binary.mavericks" on Mac OS X. If the "Built" line contains darwin13, then its mavericks, otherwise not.  And it seems that for Windows its always "windows.binary", so no reason to even check the "Built:" line.  But I don't have a windows box to test on.  Are there any other choices on mac or win?  If not, I can submit a pull request shortly.

**Comment by eddelbuettel** [comment]
@mjones:  Awesome.   Was just working on "the one other package I need to get to" before drat.

Agreed on the approach.  Best might be a help function which extracts in a tempdir and returns last line of DESCRIPTION and removes tempdir.   We can then add another helper to parse the line as well as the proper `switch()` about where to push the binary.

**Comment by mbjones** [comment]
@eddelbuettel I tested `drat::insertPackage()` with the merged PR #33 with all versions of the RNetCDF package (source, windows binary, macosx binary, and macosx mavericks binary), and they all seem to work properly and end up in the right directories.  I also tested the mavericks build for our PKIplus package and it seems to be working as well.  So I think issue #20 can be closed. Thanks for merging it.
